File: macos/QMK Toolbox/USB/Bootloader/BootloaderDevice.m
#import "BootloaderDevice.h"

@implementation BootloaderDevice

@synthesize service;

@synthesize manufacturerString;

@synthesize productString;

@synthesize vendorID;

@synthesize productID;

@synthesize revisionBCD;

- (id)initWithUSBDevice:(USBDevice *)usbDevice {
    if (self = [super init]) {
        self.usbDevice = usbDevice;
    }
    return self;
}

- (io_service_t)service {
    return self.usbDevice.service;
}

- (NSString *)manufacturerString {
    return self.usbDevice.manufacturerString;
}

- (NSString *)productString {
    return self.usbDevice.productString;
}

- (ushort)vendorID {
    return self.usbDevice.vendorID;
}

- (ushort)productID {
    return self.usbDevice.productID;
}

- (ushort)revisionBCD {
    return self.usbDevice.revisionBCD;
}

- (NSString *)description {
    return [self.usbDevice description];
}

- (void)flashWithMCU:(NSString *)mcu file:(NSString *)file {
    [NSException raise:NSInternalInconsistencyException format:@"You must override %@ in a subclass", NSStringFromSelector(_cmd)];
}

- (void)flashEEPROMWithMCU:(NSString *)mcu file:(NSString *)file {
    [NSException raise:NSInternalInconsistencyException format:@"You must override %@ in a subclass", NSStringFromSelector(_cmd)];
}

- (void)resetWithMCU:(NSString *)mcu {
    [NSException raise:NSInternalInconsistencyException format:@"You must override %@ in a subclass", NSStringFromSelector(_cmd)];
}

- (void)runProcess:(NSString *)command withArgs:(NSArray<NSString *> *)args {
    [self printMessage:[NSString stringWithFormat:@"%@ %@", command, [args componentsJoinedByString:@" "]] withType:MessageTypeCommand];

    NSTask *task = [[NSTask alloc] init];
    task.executableURL = [[NSBundle mainBundle] URLForResource:command withExtension:@""];
    task.currentDirectoryURL = [[NSBundle mainBundle] resourceURL];
    task.arguments = args;

    NSPipe *outPipe = [NSPipe pipe];
    task.standardOutput = outPipe;
    NSPipe *errPipe = [NSPipe pipe];
    task.standardError = errPipe;

    dispatch_semaphore_t semaphore = dispatch_semaphore_create(0);

    outPipe.fileHandleForReading.readabilityHandler = ^(NSFileHandle *handle) {
        NSData *data = [handle readDataOfLength:NSUIntegerMax];
        if (data.length == 0) {
            handle.readabilityHandler = nil;
            dispatch_semaphore_signal(semaphore);
        } else {
            NSString *str = [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
            [self printOutput:str];
        }
    };
    errPipe.fileHandleForReading.readabilityHandler = ^(NSFileHandle *handle) {
        NSData *data = [handle readDataOfLength:NSUIntegerMax];
        if (data.length == 0) {
            handle.readabilityHandler = nil;
            dispatch_semaphore_signal(semaphore);
        } else {
            NSString *str = [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
            [self printErrorOutput:str];
        }
    };

    [task launchAndReturnError:nil];

    dispatch_semaphore_wait(semaphore, DISPATCH_TIME_FOREVER);
    dispatch_semaphore_wait(semaphore, DISPATCH_TIME_FOREVER);
}

- (void)printOutput:(NSString *)output {
    [self printMessage:output withType:MessageTypeCommandOutput];
}

- (void)printErrorOutput:(NSString *)output {
    [self printMessage:output withType:MessageTypeCommandError];
}

- (void)printMessage:(NSString *)message withType:(MessageType)type {
    [self.delegate bootloaderDevice:self didReceiveCommandOutput:message messageType:type];
}

- (NSString *)findSerialPort {
    CFMutableDictionaryRef serialMatcher = IOServiceMatching(kIOSerialBSDServiceValue);
    CFDictionarySetValue(serialMatcher, CFSTR(kIOSerialBSDTypeKey), CFSTR(kIOSerialBSDAllTypes));

    io_iterator_t serialIterator;
    IOServiceGetMatchingServices(kIOMasterPortDefault, serialMatcher, &serialIterator);

    io_service_t port;
    while ((port = IOIteratorNext(serialIterator))) {
        ushort parentVendorID = [self.usbDevice vendorIDForService:port];
        ushort parentProductID = [self.usbDevice productIDForService:port];

        if (parentVendorID == self.vendorID && parentProductID == self.productID) {
            return [self.usbDevice stringProperty:@kIOCalloutDeviceKey forService:port];
        }
    }
    return nil;
}

@end

File: macos/QMK Toolbox/USB/Bootloader/BootloaderDevice.swift
import Foundation
import IOKit.serial

protocol BootloaderDeviceDelegate: AnyObject {
    func bootloaderDevice(_ device: BootloaderDevice, didReceiveCommandOutput data: String, type: MessageType)
}

class BootloaderDevice: USBDeviceProtocol, CustomStringConvertible {
    weak var delegate: BootloaderDeviceDelegate?

    let usbDevice: USBDevice

    var name: String = ""

    var type: BootloaderType = .none

    var eepromFlashable: Bool = false

    var resettable: Bool = false

    var service: io_service_t {
        usbDevice.service
    }

    var manufacturer: String? {
        usbDevice.manufacturer
    }

    var product: String? {
        usbDevice.product
    }

    var vendorID: UInt16 {
        usbDevice.vendorID
    }

    var productID: UInt16 {
        usbDevice.productID
    }

    var revisionBCD: UInt16 {
        usbDevice.revisionBCD
    }

    init(usbDevice: USBDevice) {
        self.usbDevice = usbDevice
    }

    func flash(_ mcu: String, file: String) {}

    func flashEEPROM(_ mcu: String, file: String) {}

    func reset(_ mcu: String) {}

    func runProcess(_ command: String, args: [String]) {
        print(message: "\(command) \(args.joined(separator: " "))", type: .command)

        let task = Process()
        task.executableURL = Bundle.main.url(forResource: command, withExtension: nil)
        task.currentDirectoryURL = Bundle.main.resourceURL
        task.arguments = args

        let outPipe = Pipe()
        task.standardOutput = outPipe
        let outHandle = outPipe.fileHandleForReading
        outHandle.waitForDataInBackgroundAndNotify()
        let errPipe = Pipe()
        task.standardError = errPipe
        let errHandle = errPipe.fileHandleForReading
        errHandle.waitForDataInBackgroundAndNotify()

        var outObserver: NSObjectProtocol!
        outObserver = NotificationCenter.default.addObserver(forName: .NSFileHandleDataAvailable, object: outHandle, queue: nil) { notification -> Void in
            let data = outHandle.availableData

            guard data.count > 0 else {
                NotificationCenter.default.removeObserver(outObserver!)
                return
            }

            self.printOutput(String(decoding: data, as: UTF8.self))
            outHandle.waitForDataInBackgroundAndNotify()
        }

        var errObserver: NSObjectProtocol!
        errObserver = NotificationCenter.default.addObserver(forName: .NSFileHandleDataAvailable, object: errHandle, queue: nil) { notification -> Void in
            let data = errHandle.availableData

            guard data.count > 0 else {
                NotificationCenter.default.removeObserver(errObserver!)
                return
            }

            self.printErrorOutput(String(decoding: data, as: UTF8.self))
            errHandle.waitForDataInBackgroundAndNotify()
        }

        do {
            try task.run()
            task.waitUntilExit()
        } catch let error {
            print(message: error.localizedDescription, type: .error)
            return
        }
    }

    func printOutput(_ output: String) {
        print(message: output, type: .commandOutput)
    }

    func printErrorOutput(_ output: String) {
        print(message: output, type: .commandError)
    }

    func print(message: String, type: MessageType) {
        delegate?.bootloaderDevice(self, didReceiveCommandOutput: message, type: type)
    }

    var description: String {
        usbDevice.description
    }

    func findSerialPort() -> String? {
        let serialMatcher = IOServiceMatching(kIOSerialBSDServiceValue)
        var serialIterator: io_iterator_t = 0

        guard IOServiceGetMatchingServices(kIOMasterPortDefault, serialMatcher, &serialIterator) == KERN_SUCCESS else { return nil }

        repeat {
            let port = IOIteratorNext(serialIterator)
            guard port != 0 else { break }

            let parentVendorID = usbDevice.vendorID(service: port)
            let parentProductID = usbDevice.productID(service: port)

            if parentVendorID == vendorID && parentProductID == productID {
                return usbDevice.stringProperty(kIOCalloutDeviceKey, service: port)
            }
        } while true

        return nil
    }
}
